### PR TITLE
add EXTERNAL_CAMP_SOURCE_DIR option to use camp from source outside RAJA

### PR DESCRIPTION
**Pull request by trws**
At the request of a code team, make an un-built external camp usable.  For an external but installed camp, add the prefix to module-path, use the same prefix, or set camp_DIR to the location of the installed campConfig.cmake.

### CODE REVIEW

**Comment by lee-busby** [comment]
I would appreciate a comment that gives the intended usage.  I see two variables that might affect how raja finds camp.  One is *EXTERNAL_CAMP_SOURCE_DIR*.  That name suggests to me the cloned camp repository directory, not an installed copy of camp.  The other variable is *camp_DIR*, which is used as the right hand side to assign a value to *EXTERNAL_CAMP_SOURCE_DIR* in one leg of the cmake conditional code.

I cloned the camp repository, ran

    cmake -DCMAKE_INSTALL_PREFIX:PATH=$HOME/tmp/camp.install .
    make; make install

This created a ``camp.install`` directory, containing an ``include`` and ``lib`` subdirectory.  The camp source directory (as cloned) has no ``lib`` subdirectory.

I want to point my raja build at the ``$HOME/tmp/camp.install`` directory.  Do I say ``-Dcamp_DIR=$HOME/tmp/camp.install`` to do that?  Sorry to be dense.

**Comment by trws** [comment]
To use an installed copy you can either use camp_DIR=install-path/lib or add the path as a prefix or the lib directory to the module path list. The new one makes it so you can use an external clone of camp.

Sent from my iPhone

On Dec 18, 2019, at 11:46 AM, lee-busby <notifications@github.com> wrote:

﻿

I would appreciate a comment that gives the intended usage. I see two variables that might affect how raja finds camp. One is EXTERNAL_CAMP_SOURCE_DIR. That name suggests to me the cloned camp repository directory, not an installed copy of camp. The other variable is camp_DIR, which is used as the right hand side to assign a value to EXTERNAL_CAMP_SOURCE_DIR in one leg of the cmake conditional code.

I cloned the camp repository, ran

cmake -DCMAKE_INSTALL_PREFIX:PATH=$HOME/tmp/camp.install .
make; make install


This created a camp.install directory, containing an include and lib subdirectory. The camp source directory (as cloned) has no lib subdirectory.

I want to point my raja build at the $HOME/tmp/camp.install directory. Do I say -Dcamp_DIR=$HOME/tmp/camp.install to do that? Sorry to be dense.

**Comment by lee-busby** [comment]
Ok, that sounds good - thank you.

File: CMakeLists.txt [hunk 1]
```diff
@@ -201,11 +201,21 @@ if (ENABLE_TBB)
     tbb)
 endif ()
 
-find_package(camp QUIET)
-if (NOT camp_FOUND)
-  message(STATUS "Using RAJA CAMP submodule.")
-  add_subdirectory(tpl/camp)
-endif(NOT camp_FOUND)
+set(EXTERNAL_CAMP_SOURCE_DIR "" CACHE FILEPATH "Build OpenMP support")
```
[the remark this answers is not in the document]
Comment (trws) [reply CMakeLists.txt]: Thanks, how on earth I missed that will remain a mystery... 🤦‍♂

**Comment by trws** [comment]
I put in a small tweak to get more meaningful output from the camp detection.  This can be used to build either with an external source tree, or if you want to use an installed version, it works like this:

`cmake -DCMAKE_PREFIX_PATH=<camp-install-path> <other-cmake-options>`

If there are things you don't want cmake to use in the same path for some reason, you can also use `camp_DIR` like this:

`cmake -Dcamp_DIR=<camp-install-path>/lib/cmake/camp <other-cmake-options>`

That will use *only* camp from that path and leave everything else pointing the normal places.

It's also rebased, anyone want to give it a poke before we go forward with it?

**Pull request merged**